lapce-ui/src/problem.rs

```diff
@@ -94,89 +92,111 @@ impl ProblemContent {
         mouse_event: &MouseEvent,
         data: &LapceTabData,
     ) {
-        let n = (mouse_event.pos.y / self.line_height).floor() as usize;
+        let click_line = (mouse_event.pos.y / self.line_height).floor() as usize;
 
         let items = self.items(data);
-        let mut i = 0;
-        for (path, diagnostics) in items {
-            let diagnostics_len = diagnostics.iter().map(|d| d.lines).sum::<usize>();
-            if diagnostics_len + 1 + i < n {
-                i += diagnostics_len + 1;
-                continue;
+        let mut line_cursor = 0;
+
+        let mut it = items.into_iter().peekable();
+
+        while let Some((path, diagnostics)) = it.peek() {
+            let is_collapsed = *data.problem.fold.get(*path).unwrap_or(&false);
+            let offset = if is_collapsed {
+                1
+            } else {
+                let diagnostics_len =
+                    diagnostics.iter().map(|d| d.lines).sum::<usize>();
+                diagnostics_len + 1
+            };
+            if offset + line_cursor >= click_line {
+                break;
             }
+            line_cursor += offset;
+            it.next();
+        }
 
-            for d in diagnostics {
-                if i > n {
+        for (path, diagnostics) in it {
+            if click_line == 0 || line_cursor == click_line {
+                ctx.submit_command(Command::new(
+                    LAPCE_UI_COMMAND,
+                    LapceUICommand::ToggleProblem(path.to_path_buf()),
+                    Target::Widget(data.id),
+                ));
+                return;
+            }
+
+            for file_diagnostic in diagnostics {
+                if line_cursor > click_line {
                     return;
                 }
 
-                let msg_lines = d.diagnostic.message.matches('\n').count() + 1;
-                let related_lines = d
```

Comment:
We need to count related_lines as well.

Reply:
They are already counted in `EditorDiagnostic` `lines` as I wrote in docs of this field (with English so slaughted I can't believe I wrote it so I need to fix at least this part).

I tested this couple times in various ways (1 file with multiple warnings, 2 files, 3 files, toggle different files and clicked on different warnings) and it seems to be working correctly.

Reply:
Docs fixed and some additional where added